boudicca.base/enricher/src/main/kotlin/base/boudicca/enricher/service/osm/OsmEnricher.kt

```diff
@@ -0,0 +1,218 @@
+package base.boudicca.enricher.service.osm
+
+import base.boudicca.Property
+import base.boudicca.SemanticKeys
+import base.boudicca.enricher.service.Enricher
+import base.boudicca.enricher.service.EnricherOrderConstants
+import base.boudicca.fetcher.Fetcher
+import base.boudicca.fetcher.InMemoryFetcherCache
+import base.boudicca.format.ListFormat
+import base.boudicca.format.UrlUtils
+import base.boudicca.format.UrlUtils.encodeURL
+import base.boudicca.model.structured.Key
+import base.boudicca.model.structured.StructuredEvent
+import base.boudicca.model.structured.StructuredEvent.StructuredEventBuilder
+import base.boudicca.model.structured.Variant
+import base.boudicca.model.structured.VariantConstants.SOURCE_VARIANT_NAME
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import org.slf4j.LoggerFactory
+import org.springframework.core.annotation.Order
+import org.springframework.stereotype.Service
+
+
+private const val NOMINATIM_BASE_URL = "https://nominatim.boudicca.events"
+
+@Service
+@Order(EnricherOrderConstants.OsmEnricherOrder)
+class OsmEnricher(
+    private val fetcher: Fetcher = createDefaultFetcher()
+) : Enricher {
+    val sourceVariant = listOf(Variant(SOURCE_VARIANT_NAME, "OSM"))
+    val logger = LoggerFactory.getLogger(this::class.java)!!
+
+    val mapper = jacksonObjectMapper()
+
+    override fun enrich(event: StructuredEvent): StructuredEvent {
+        val osmId: String? = getOrSearchOsmId(event)
+        if (osmId == null || !isValidOsmIdFormat(osmId)) {
+            return event
+        }
+        val enrichedEvent = enrichEventFromOverpass(osmId, event)
+
+        return enrichedEvent
+    }
+
+    @Suppress("detekt.LongMethod")
+    private fun enrichEventFromOverpass(
+        osmId: String,
+        event: StructuredEvent
+    ): StructuredEvent {
+        val url = "$NOMINATIM_BASE_URL/lookup?format=jsonv2&extratags=1&namedetails=1&osm_ids=${osmId}"
+        val response = synchronized(fetcher) {
+            fetcher.fetchUrl(url)
+        }
+
+        // parse osm data into our properties
+        val builder = event.toBuilder()
+        val json = mapper.readTree(response)
+        if (json.isEmpty) {
+            logger.warn("nominatim did not return any results for osm id $osmId")
+            return event
+        }
+        val firstElement = json.first()
```

Comment:
would `rootElement` or `nominatimLocationJson` or `locationJson` or `location` or something maybe be better names than firstElement? i'm guessing the first is because there might be multiple results? should there be multiple results, should we use single() instead of first()?

Reply:
i renamed it to `nominatimPlace` because the structure is a "place" as described in https://nominatim.org/release-docs/develop/api/Output/

also since we lookup a single osmId there can only be a single result (or none), so i will switch it to `single()`